Bluetooth: Port mesh shell to the new shell

Vudentz opened the pull request:
This port the mesh shell which was still using the legacy shell to the new shell.

Vudentz committed:
shell: Add APIs to set number of arguments

This allows the shell core to perform precheck before calling the
handler which then can assume the number of arguments is correct.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
mesh: Port shell module to the new shell

This makes mesh shell to register commands with the new shell subsystem
and stop using the legacy shell.

Fixes #11056

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

jhedberg reviewed (approved):
Looks generally good. In the long run we might want to refactor these into sub commands.

jarz-nordic commented on include/shell/shell.h:
```diff
@@ -71,6 +71,22 @@ struct shell_cmd_entry {
+	.optional = _optional,						\
+}
+
+struct shell_static_args {
```
Argument `argc` in command handler is `size_t` type, just a proposal to change these parameters to `size_t` as well.

Vudentz replied:
Initially I was using that but it takes a lot of space, besides size_t + size_t may overflow (not that I think someone will be crazy enough to have this manu arguments).

jarz-nordic commented on include/shell/shell.h:
```diff
@@ -97,7 +114,32 @@ struct shell_static_entry {
  * @param[in] handler Pointer to a function handler.
  */
-#define SHELL_CMD_REGISTER(syntax, subcmd, help, handler)		   \
+#define SHELL_CMD_ARG_REGISTER(syntax, subcmd, help, handler,		   \
```
I would update a bit macro description to exaplain difference between `SHELL_CMD_REGISTER` and `SHELL_CMD_ARG_REGISTER`

Vudentz replied:
Ok

jakub-uC reviewed (changes requested):
Great improvement!

Please add one more commit and update paragraph: `Creating commands` in shell.rst with new macro: `SHELL_CMD_ARG_REGISTER`

Vudentz commented on subsys/bluetooth/host/mesh/Kconfig:
```diff
@@ -425,7 +425,8 @@ config BT_MESH_HEALTH_CLI
 config BT_MESH_SHELL
 	bool "Enable Bluetooth Mesh shell"
-	select CONSOLE_SHELL
+	select SERIAL
```
And where is that selected?

jarz-nordic replied:
Kconfig.backends->config SHELL_BACKEND_SERIAL

Vudentz commented on subsys/bluetooth/host/mesh/shell.c:
```diff
@@ -28,6 +28,15 @@
+static const struct shell *ctx_shell;
+
+#define print(_sh, _ft, ...) \
+	shell_fprintf(_sh ? _sh : ctx_shell, SHELL_NORMAL, _ft "\r\n", \
```
Alright, seems that has changed since the last time I did the port for other shells.

Vudentz commented on subsys/shell/shell.c:
```diff
@@ -946,6 +946,52 @@ static const struct shell_cmd_entry *root_cmd_find(const char *syntax)
+		if (args->optional > 0) {
+			/* Check if argc is within allowed range */
+			ret_val = shell_cmd_precheck(shell,
+						     (argc >= args->mandatory &&
```
Not sure where though, the expression seems to be fine given our code style and checkpatch doesn't complain either.

jarz-nordic replied:
It's only my preference not a must be change :)

```
ret_val = shell_cmd_precheck(shell,
                             (argc >= args->mandatory) &&
                             (argc <= (args->mandatory +
                                       args->optional)),
                             NULL, 0);
```

Vudentz replied:
Right, I though by bracket you meant square brackets like in [], round () are more common refereed as parentheses, or perhaps I assume so giver that brackets can be used for more than one thing. Anyway I will add the extra parantheses/round brackets.

jarz-nordic commented on tests/bluetooth/mesh_shell/src/main.c:
```diff
@@ -11,12 +11,11 @@
 
-#include <shell/legacy_shell.h>
+#include <shell/shell.h>
+#include <shell/shell_uart.h>
```
This is not needed anymore. 
Backend is selected by Kconfig and its initialization is realized in `shell_uart.c` you do not need to do it in main.c file.

Vudentz committed:
shell: Add macros to print messages

This adds macros for printing info, normal, warning and error messages
including line termination.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz commented:
recheck

jarz-nordic commented:
@Vudentz : Please remember to update shell documentation with information about  macro `SHELL_CMD_ARG_REGISTER` in this place:

(doc/subsys/shell/shell.rst)   paragraph: Creating commands

dbkinder commented on include/shell/shell.h:
```diff
@@ -497,6 +557,54 @@ int shell_stop(const struct shell *shell);
+ *
+ * This function shall not be used outside of the shell command context.
+ *
+ * @param[in] shell Pointer to the shell instance.
```
parameter names in the doxygen comments must match actual names: ``_sh`` and ``_ft``
